dcc-portal-ui/app/scripts/pathwayviewer/js/pathways.js

```diff
@@ -209,149 +214,199 @@
             _zooms = [''];
             deferred.resolve();
           }
-          
+
           return deferred.promise;
         })
         .then(function () {
           var deferred = $q.defer();
-          
-          
+
           GeneSetService.getPathwayProteinMap(_parentPathwayId, []).then(function (map) {
-            
+ 
             // Normalize into array
             _.forEach(map, function (value, id) {
-              
               if (value && value.dbIds) {
                 var dbIds = value.dbIds.split(',');
-                
+
                 _pathwayMutationHighlights.push({
                   uniprotId: id,
                   dbIds: dbIds,
                   value: value.value
                 });
               }
             });
-            
-            //console.log(_geneOverlapExistsHash);
-            // Get ensembl ids for all the genes so we can link to advSearch page
-            _uniprotIds = _.pluck(_pathwayMutationHighlights, 'uniprotId');
-            deferred.resolve();
-          });
-          return deferred.promise;
-        })
-        .then(function () {
-          var deferred = $q.defer();
-          
-          Restangular.one('genes').get({filters: pathway.geneSetOverlapFilters})
-            .then(function (geneListData) {
-              
+
+            SetService.getTransientSet('GENE', {
+              'filters': {gene:{pathwayId:{is:[_parentPathwayId]}}},
+              'sortBy': 'id',
+              'sortOrder': 'ASCENDING',
+              'name': _parentPathwayId,
+              'size': _geneSet.geneCount,
+              'transient': true 
+            }).then(function (entitySetData) {
+              CompoundsService.getCompoundsFromEntitySetId(entitySetData.id).then(function(drugs) {                
+                var drugMap = {};
+                _.forEach(drugs, function(drug) {
+                  _.forEach(drug.genes, function(gene) {
+                    var uniprotId = gene.uniprot;
+
+                    if (_.isUndefined(drugMap[uniprotId])) {
+                      drugMap[uniprotId] = [];
+                    }
+
+                    drugMap[uniprotId].push({
+                      'zincId': drug.zincId,
+                      'name': drug.name
+                    });
+                  });
+                });
+
+                // Normalize into array
+                _.forEach(map,function(value,id) {
+                  if(value && value.dbIds && drugMap[id]) {
+                    _pathwayDrugHighlights.push({
+                      uniprotId:id,
+                      dbIds:value.dbIds.split(','),
+                      drugs:drugMap[id]
+                    });
+                  }
+                });
+                deferred.resolve();
+              });
+              //console.log(_geneOverlapExistsHash);
+            });
+            return deferred.promise;
+          })
+          .then(function () {
+            var deferred = $q.defer();
```

Comment:
There's a [promise antipattern](https://github.com/petkaantonov/bluebird/wiki/Promise-anti-patterns) where deferred are being created and promise chains broken up for no reason, and that's being done throughout this file. I see that this isn't your fault and has been in the code for some time now.

I think it would be worthwhile to clean this up. @btiernay what do you think?

Poke me next time you're in our corner of the building and I can help you out if you have trouble with it